cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
+        for (std::int64_t i = 0; i < size_; i++) {
+            (*this)[i] = element;
+        }
+    }
 
     template <typename Deleter>
     explicit array(T* data, std::int64_t size, Deleter&& deleter)
-        : data_(data, deleter),
-          size_(size) {}
+        : data_owned_ptr_(data, std::forward<Deleter>(deleter)),
+          data_(data),
+          size_(size),
+          capacity_(size) {}
 
     template <typename U>
     array(const array<U>& other)
-        : data_(other.data_),
-          size_(other.size_) {}
+        : data_owned_ptr_(other.data_owned_ptr_),
+          data_(other.data_),
+          size_(other.size_),
+          capacity_(other.capacity_) {}
+
+    T* get_mutable_data() const {
+        return std::get<T*>(data_);
+    }
+
+    const T* get_data() const {
+        if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val;
+        } else {
+            return std::get<const T*>(data_);
+        }
+    }
 
-    T* get_pointer() const noexcept {
-        return data_.get();
+    bool has_mutable_data() const {
+        return std::holds_alternative<T*>(data_) && (get_mutable_data() != nullptr);
     }
 
-    std::int64_t get_size() const noexcept {
+    void unique() {
+        if (is_data_owner() || size_ == 0) {
+            return;
+        } else {
+            auto immutable_data = get_data();
+            auto copy_data = new T[size_];
+
+            for (std::int64_t i = 0; i < size_; i++) {
+                copy_data[i] = immutable_data[i];
+            }
+
+            reset(copy_data, size_, [](auto obj){ delete[] obj; });
+        }
+    }
+
+    std::int64_t get_size() const {
         return size_;
     }
 
+    std::int64_t get_capacity() const {
+        return capacity_;
+    }
+
+    bool is_data_owner() const {
+        if (data_owned_ptr_ == nullptr) {
+            return false;
+        } else if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else if (auto ptr_val = std::get_if<const T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else {
+            return false;
+        }
+    }
+
     void reset() {
-        data_.reset();
+        data_owned_ptr_.reset();
+        data_ = std::variant<T*, const T*>();
         size_ = 0;
+        capacity_ = 0;
+    }
+
+    void reset(std::int64_t size) {
+        data_owned_ptr_.reset(new T[size], [](auto obj){ delete[] obj; });
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
+    }
+
+    template <typename Deleter>
+    void reset(T* data, std::int64_t size, Deleter&& deleter) {
+        // TODO: check input parameters
+        data_owned_ptr_.reset(data, std::forward<Deleter>(deleter));
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
+    }
+
+    template <typename U = T*>
+    void reset_not_owning(U data = nullptr, std::int64_t size = 0) {
```

Comment:
If we decide to apply [this comment](https://github.com/rlnx/daal/pull/4/files?file-filters%5B%5D=.hpp#r419109312), we wouldn't need this one.

Reply:
I do not see how `resize()` method with proposed corrections will help. There is also a case when user wants to forget data he owns and resets an `array` to another (but immutable) data. This method is the only way to do this for now since when all the constructors (except default ctor) create mutable owning arrays.

Reply:
>There is also a case when user wants to forget data he owns and resets an array to another (but immutable) data

The problem is that the method does not "forget" the pointer it owns. Is that bug in implementation?

Reply:
Ok, I finally got what this method is doing.

Reply:
Need to look for a better name, I had not got the meaning from the name, so I've written the comment.